Add Matter update entities for devices with OTA requestor

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Matter devices which support the OTA requestor cluster can receive updates from a OTA provider. The Home Assistant Python Matter Server implements such an OTA provider now.

Add update entities for devices which support the OTA requestor cluster and check for available updates. Allow the user to update the firmware. The update progress will be read directly from the devices' OTA requestor cluster.

Related Python Matter Server PR: https://github.com/home-assistant-libs/python-matter-server/pull/709


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33874

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

agners committed:
Add Matter update entities for devices with OTA requestor

Matter devices which support the OTA requestor cluster can receive
updates from a OTA provider. The Home Assistant Python Matter Server
implements such an OTA provider now.

Add update entities for devices which support the OTA requestor cluster
and check for available updates. Allow the user to update the firmware.
The update progress will be read directly from the devices' OTA
requestor cluster.

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+from .helpers import get_matter
+from .models import MatterDiscoverySchema
+
+SCAN_INTERVAL = timedelta(hours=12)
```
I'm surprised by this combined with `should_poll` not being overridden in the update entity class.
Is the `matter` integration polled?

agners replied:
No, the Matter integration is not polled.

But checking for updates is. The device don't have a bulit-in update checking mechanism, so we don't get a notification pushed when there is an update.

We *could* do the polling on Matter Server side. But it feels more naturally to use Core's polling capability. With this, one can also use the `homeassistant.update_entity` service to force an update check.

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+            self._attr_in_progress = True
+
+    async def async_update(self) -> None:
+        """Call when the entity needs to be updated."""
```
```suggestion
        """Update the state of the entity."""
```

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+        else:
+            self._attr_in_progress = True
+
+    async def async_update(self) -> None:
```
Why does this method not just call `self._update_from_device` ?

agners replied:
`_update_from_device` gets called on every attribute update as pushed by the device. We don't want to trigger a update check whenever a random attribute changes on the device.

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+        """Return Matter specific state data to be restored."""
+        return MatterUpdateExtraStoredData(self._software_update)
+
+    @property
+    def entity_picture(self) -> str | None:
+        """Return the entity picture to use in the frontend."""
+
+        # The Matter brand picture is not appropriate as this is the update
+        # entity for the device.
+        return None
```
What sets the entity picture in other matter entities?

agners replied:
It is not being set in other places. This is more about overwriting the default behavior of the Update entity.

emontnemery replied:
Ok, then please rewrite the docstring to make that clearer.
Maybe like this + remove the comment:
```py
    def entity_picture(self) -> str | None:
        """Return the entity picture to use in the frontend.

        This overrides UpdateEntity.entity_picture because the Matter brand picture
        is not appropriate for a matter device which has its own brand.
        """
```

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+            raise HomeAssistantError("No software version specified")
+
+        self._attr_in_progress = True
+        self.async_write_ha_state()
```
Why is this needed? If the entity is polled, an update should be scheduled after service calls

agners replied:
It usually takes 2-3s for the device to get the first update progress state changes. Pushing the progress state change here, really improves user expirience: With that, the frontend immeaditly switches into a "in progress" animation. Without it, the frontend doesn't update, and it feels like the "Install" button is broken.

agners replied:
It usually takes 2-3s for the device to get the first update progress state changes. Pushing the progress state change here, really improves user expirience: With that, the frontend immeaditly switches into a "in progress" animation. Without it, the frontend doesn't update, and it feels like the "Install" button is broken.

emontnemery replied:
OK, please add a comment explaining we immediately update the state to not have to wait for the real update state change.

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,221 @@
+                self.hass, POLL_AFTER_INSTALL, self._async_update_future
+            )
+
+    async def _async_update_future(self, now: datetime | None = None) -> None:
+        """Request update."""
+        await self.async_update()
+
```
This is not needed

```suggestion
```

TheJulianJES commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,225 @@
+
+@dataclass
+class MatterUpdateExtraStoredData(ExtraStoredData):
+    """Extra stored data for Z-Wave node firmware update entity."""
```
```suggestion
    """Extra stored data for Matter node firmware update entity."""
```

agners committed:
Update homeassistant/components/matter/update.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

agners committed:
Bump python-matter-server to 6.3.0

This includes models and commands required for device firmware updates.

marcelveldt reviewed (approved):
From a Matter POV this one is approved - I'll leave a final approval to Erik as member of the core team.

One note is that we agreed to show the extra warning, for which an architecture discussion is created and an implementation (at core level) needs to be implemented. 

If we agree to merge this now, we should also make sure to fix that dialog before the release of HA 2024.8

agners committed:
Fix tests by including the new bluetooth_enabled field

agners committed:
Add update entity tests

TheJulianJES commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,225 @@
+    async def async_update(self) -> None:
+        """Call when the entity needs to be updated."""
+        try:
+            update_information = await self.matter_client.check_node_update(
```
Does this send a message to the Matter devices too or is it just a local update check?

If the update button on the "Updates" page or the "Check for updates" menu button on the HA settings page is clicked, `async_update` is called for all update entities.
If an "image notify" message is sent to each device (e.g. via Thread) here, this could cause the network to be spammed for a short amount of time.

So, if this sends a message to the device, maybe set `PARALLEL_UPDATES` to `1` or another reasonable number for the Matter update entity/platform?

agners replied:
> Does this send a message to the Matter devices too or is it just a local update check?

This causes no communication with devices. It uses cached vendor/product ID information to check for updates. The update check is polling the http get requests to the CSA DCL REST API.

agners committed:
Fix update entity test

agners committed:
Update entity picture docstring

agners committed:
Add note about reasons for progress state change update

agners committed:
Enable polling for update entities by default

Matter entities don't enable polling any longer. Enable polling for
update entities by default.

emontnemery commented on homeassistant/components/matter/update.py:
```diff
@@ -0,0 +1,229 @@
+class MatterUpdate(MatterEntity, UpdateEntity):
+    """Representation of a Matter node capable of updating."""
+
+    _attr_should_poll = True
```
Maybe explain why the update entity is polled

agners replied:
Added. Thanks for the review!

emontnemery reviewed (approved):
LGTM

agners committed:
Add comment about why Update entities are polled

MartinHjelmare commented on tests/components/matter/test_update.py:
```diff
@@ -0,0 +1,171 @@
+        == "http://home-assistant.io/non-existing-product"
+    )
+
+    await async_setup_component(hass, "update", {})
```
Why is this needed?

agners replied:
This is not needed, I'll remove it.

MartinHjelmare reviewed (commented):
This PR was merged with only 85% test coverage. Please add more tests.

agners commented:
> This PR was merged with only 85% test coverage. Please add more tests.

PR https://github.com/home-assistant/core/pull/122575 adds more tests.
